pkg/codec/modifier_codec.go

```diff
@@ -0,0 +1,185 @@
+package codec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+)
+
+func NewModifierCodec(codec types.Codec, typeProvider types.TypeProvider, modifier Modifier) (types.Codec, error) {
+	if codec == nil || typeProvider == nil || modifier == nil {
+		return nil, errors.New("inputs must not be nil")
+	}
+
+	return &modifierCodec{
+		codec:        codec,
+		typeProvider: typeProvider,
+		modifier:     modifier,
+	}, nil
+}
+
+var _ types.TypeProvider = &modifierCodec{}
+
+type modifierCodec struct {
+	codec        types.Codec
+	typeProvider types.TypeProvider
+	modifier     Modifier
+}
+
+func (m *modifierCodec) CreateType(itemType string, forEncoding bool) (any, error) {
+	t, err := m.typeProvider.CreateType(itemType, forEncoding)
+	if err != nil {
+		return nil, err
+	}
+	ot := reflect.TypeOf(t)
+	nt, err := m.modifier.RetypeForOffChain(ot)
+	if err != nil {
+		return nil, err
+	}
+
+	return reflect.Zero(nt).Interface(), nil
+}
+
+func (m *modifierCodec) Encode(ctx context.Context, item any, itemType string) ([]byte, error) {
+	tmpOnChain, err := m.typeProvider.CreateType(itemType, true)
+	if err != nil {
+		return nil, err
+	}
+
+	offChainType, err := m.modifier.RetypeForOffChain(reflect.TypeOf(tmpOnChain))
+	if err != nil {
```

Comment:
Just to make sure I understood. `RetypeForOffChain` returns a type that should be a modified version of `item` for eg. adding a field?
What should `TransformForOffChain` do, rename fields, hard code values?

I feel like some of this code could use a bit more comments even if tests are helpful for understanding it, it can get a bit confusing unless you used reflect package recently.

Reply:
In the simplest terms, with respect to relay, `RetypeForOffChain` will change the type used on-chain to something that can be parsed from the input json.  `TransformForOffChain` does the transformation of the on-chain to the off-chain object.